_posts/en/newsletters/2019-03-05-newsletter.md

```diff
@@ -0,0 +1,469 @@
+no more than 10,000 bytes (0.01 MB).
+
+Even more work can be required if signatures (or things that pretend
+to be signatures) are included directly in a scriptPubKey as this will
+cause an internal `FindAndReplace()` operation to modify the
+executed script and again cause sigops to commit to a separate copy of
+almost the entire transaction.  Because a secure signature in Bitcoin
+commits to the scriptPubKey it spends, and because a signature in a
+scriptPubKey can't commit to itself, there's no legitimate reason to
+check a signature included in a scriptPubKey.  [BIP143][] segwit
+addresses this problem for its spends by simply specifying that
+`FindAndReplace()` shall not be used.
+
+Finally, the original Bitcoin Protocol also allows an attacker to add
+non-push opcodes to scriptSig to use up to an additional 20,000 sigops
+in a block as well as perform other operations (such as using `OP_DUP`
+(duplicate)) to increase the amount of verification work that needs to
+be done.
+
+Combining all of these problems together makes it possible for a
+well-prepared attacker to create blocks that take a long time to verify
+even on fast hardware.  Optech doesn't know how long the worst case is,
+as researchers keep their example scripts private to avoid arming
+attackers.  We've confidently heard that blocks can be created that take
+more than half an hour to verify on fast modern hardware.  (Before the
+correction of other known problems with the above mechanisms, tests show
+that it was possible to create blocks that took [several hours to
+verify][sdl findanddelete].)  An attacking miner could use these
+problems to denial-of-service attack verification nodes and other
+miners, possibly finding ways to profit from the situation.  However,
+because the attacks involve rarely-used Bitcoin features, any actual
+attack would probably be met by a soft fork to immediately disable the
+features---ensuring Bitcoin would return to normal as soon as the fork
+activated.
+
+We have no way to solve the problem of each legacy input requiring the
+hashing of a slightly different set of transaction data without
+forbidding the use of legacy transaction signatures altogether, but the consensus
+cleanup soft fork proposes to prevent the magnifications of the problem
+by forbidding use of `OP_CODESEPARATOR` in legacy inputs, the behavior
+that triggers `FindAndReplace()`, and the ability to use non-data
+pushing opcodes in scriptSig.  Many developers believe this is
+acceptable because there are no known productive uses of the behavior,
+there's no onchain activity visible of anyone using them, and because
+people who want to play with `OP_CODESEPARATOR` can still do so using
+the non-problematic version available in segwit.  Together with
+improvements in caching, this could put worst-case block verification
+into the range of seconds rather than minutes.
+
+More information:
+
+- [The Megatransaction][megatransaction] by Rusty Russel - a block that
```

Comment:
Rusty Russell with two Ls

Reply:
Thanks so much for this!  I somehow keep making this mistake, and it's embarrassing each time.